clientconn.go

```diff
@@ -133,12 +140,35 @@ func (cc *ClientConn) resetTransport(closeTransport bool) error {
 		if closeTransport {
 			t.Close()
 		}
-		newTransport, err := transport.NewClientTransport(cc.dopts.protocol, cc.target, cc.dopts.authOptions)
+		// Adjust timeout for the current try.
+		dopts := cc.dopts
+		if dopts.Timeout < 0 {
+			cc.Close()
+			return ErrClientConnTimeout
+		}
+		if dopts.Timeout > 0 {
+			dopts.Timeout -= time.Since(start)
+			if dopts.Timeout <= 0 {
+				cc.Close()
+				return ErrClientConnTimeout
+			}
+		}
+		newTransport, err := transport.NewClientTransport(cc.target, &dopts)
 		if err != nil {
-			// TODO(zhaoq): Record the error with glog.V.
+			if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
+				cc.Close()
+				return ErrClientConnTimeout
+			}
+			sleepTime := backoff(retries)
+			// Fail early before falling into sleep.
+			if dopts.Timeout > 0 && cc.dopts.Timeout < sleepTime + time.Since(start) {
```

Comment:
Ah, I meant instead of. So

if cc.dopts.Timeout > 0 && cc.dopts.Timeout < sleepTime + time.Since(start) {

instead of

if dopts.Timeout > 0 && cc.dopts.Timeout < sleepTime + time.Since(start) {



But you're right that we're constrained cc.dopts.Timeout >= dopts.Timeout. I just thought it made more sense to use the latter var, since we're explicitly checking that it was set.